File: web/src/shared/auth/permissions.ts
/**
 * Карта прав `роль → (право → boolean)`.
 *
 * `SPEC.md § 2.2`. Источник истины — сервер: на M3 карта будет приходить
 * в `/auth/me` и генерироваться из `accounts/permissions.py`. Здесь она задана
 * статически, чтобы макеты показывали навигацию так же, как будет в бою.
 *
 * **Скрытие элемента в интерфейсе не заменяет проверку на сервере.** Проверка
 * в каждом эндпоинте обязательна и возвращает 403; эта карта управляет только
 * видимостью (`SPEC.md § 2.3`).
 */
import type { Role } from '@/api/types';

export const PERMISSIONS = [
  // Операции
  'schedule.view',
  'flight.create',
  'flight.edit',
  'flight.status',
  'flight.request', // клиент создаёт заявку, а не рейс
  'request.approve',
  // Услуги
  'service.order',
  'service.confirm',
  'service.confirm.own', // поставщик подтверждает только свои
  'catalog.view',
  'catalog.edit',
  // Контрагенты
  'vendor.view',
  'vendor.edit',
  'vendor.assign',
  'contract.view',
  'contract.edit',
  'client.view',
  'client.edit',
  'tariff.view',
  'tariff.edit',
  // Финансы
  'billing.documents.view',
  'billing.documents.edit',
  'billing.payables.view',
  'billing.payables.edit',
  'billing.reconciliation',
  'billing.margin.view',
  'billing.purchase_price.view', // закупочные цены — не для клиента
  // Коммуникации
  'comms.view',
  'comms.templates.edit',
  // Отчётность
  'reports.operational',
  'reports.financial',
  'reports.vendor.own',
  'dashboard.dispatcher',
  'dashboard.manager',
  // Администрирование
  'audit.view',
  'admin',
] as const;

export type Permission = (typeof PERMISSIONS)[number];

type PermissionMap = Readonly<Record<Permission, boolean>>;

function grant(...granted: Permission[]): PermissionMap {
  const map = Object.fromEntries(PERMISSIONS.map((key) => [key, false])) as Record<
    Permission,
    boolean
  >;
  for (const key of granted) map[key] = true;
  return map;
}

/**
 * Матрица из `SPEC.md § 2.2`.
 *
 * Роль `sales` добавлена по ТЗ 2.1 (ADR-011) и подлежит подтверждению
 * заказчиком — G-06. Пока подтверждения нет, пользователи с ней не создаются,
 * но право объявлено: добавить роль в карту сейчас стоит одну строку, после
 * написания проверок в каждом эндпоинте — правку всех эндпоинтов.
 */
export const ROLE_PERMISSIONS: Readonly<Record<Role, PermissionMap>> = {
  admin: Object.fromEntries(PERMISSIONS.map((key) => [key, true])) as PermissionMap,

  dispatcher: grant(
    'schedule.view',
    'flight.create',
    'flight.edit',
    'flight.status',
    'request.approve',
    'service.order',
    'service.confirm',
    'catalog.view',
    'vendor.view',
    'vendor.assign',
    'contract.view',
    'client.view',
    'billing.documents.view',
    'billing.margin.view',
    'billing.purchase_price.view',
    'comms.view',
    'reports.operational',
    'dashboard.dispatcher',
  ),

  sales: grant(
    'schedule.view',
    'client.view',
    'tariff.view',
    'catalog.view',
    'billing.documents.view',
    'billing.documents.edit',
    'comms.view',
    'reports.financial',
  ),

  finance: grant(
    'schedule.view',
    'catalog.view',
    'catalog.edit',
    'vendor.view',
    'contract.view',
    'contract.edit',
    'client.view',
    'client.edit',
    'tariff.view',
    'tariff.edit',
    'billing.documents.view',
    'billing.documents.edit',
    'billing.payables.view',
    'billing.payables.edit',
    'billing.reconciliation',
    'billing.margin.view',
    'billing.purchase_price.view',
    'comms.view',
    'reports.financial',
  ),

  manager: grant(
    'schedule.view',
    'catalog.view',
    'vendor.view',
    'contract.view',
    'client.view',
    'tariff.view',
    'billing.documents.view',
    'billing.payables.view',
    'billing.margin.view',
    'billing.purchase_price.view',
    'comms.view',
    'reports.operational',
    'reports.financial',
    'dashboard.dispatcher',
    'dashboard.manager',
    'audit.view',
  ),

  // Портал клиента: свои рейсы и свои документы. Закупочных цен и данных
  // поставщиков нет по построению ответа, а не по фильтрации на клиенте.
  client: grant('schedule.view', 'flight.request', 'billing.documents.view'),

  // Портал поставщика: только свои заявки и свои показатели.
  vendor: grant('service.confirm.own', 'billing.payables.view', 'reports.vendor.own'),
};

export function hasPermission(role: Role, permission: Permission): boolean {
  return ROLE_PERMISSIONS[role][permission];
}

File: backend/accounts/permissions.py
"""Карта прав `[ТЗ 4.3]`.

Единый источник истины с клиентами (`BACKEND.md § 6`): карта отдаётся
в `/auth/me`, веб-клиент и мобильное приложение по ней прячут элементы
интерфейса. Прятать — не значит запрещать: проверка на сервере обязательна
в каждом эндпоинте, иначе достаточно открыть консоль браузера.

Право отвечает на вопрос «можно ли вообще», а не «к чьим записям».
Ограничение «только свои» из `SPEC.md § 2.2` — это не право, а фильтрация
выборки: она живёт в `core.api.viewsets.TenantScopedViewSet` и работает
в `get_queryset()`. Смешивать их нельзя: право, выданное с оговоркой
«к своим», в списочном эндпоинте протекает.
"""

from __future__ import annotations

from django.db import models
from django.utils.translation import gettext_lazy as _

from accounts.models import Role


class Permission(models.TextChoices):
    """Разрешения. Разбиение — по разделам матрицы `SPEC.md § 2.2`."""

    FLIGHTS_READ = "flights.read", _("Суточный план, чтение")
    FLIGHTS_WRITE = "flights.write", _("Создание и изменение рейса")
    FLIGHTS_STATUS = "flights.status", _("Смена статуса рейса")

    # Клиент не создаёт рейс напрямую: он подаёт заявку, диспетчер
    # подтверждает её, и только тогда появляется рейс (SPEC § 2.2, сноска).
    FLIGHT_REQUESTS_CREATE = "flight_requests.create", _("Подача заявки на рейс")
    FLIGHT_REQUESTS_REVIEW = "flight_requests.review", _("Рассмотрение заявок на рейс")

    ORDERS_CREATE = "orders.create", _("Заказ услуг")
    ORDERS_REQUEST = "orders.request", _("Запрос услуги заявкой")
    ORDERS_CONFIRM = "orders.confirm", _("Подтверждение услуги")

    CATALOG_READ = "catalog.read", _("Каталог услуг и цены, чтение")
    CATALOG_WRITE = "catalog.write", _("Каталог услуг и цены, изменение")

    VENDORS_READ = "vendors.read", _("Реестр поставщиков, чтение")
    VENDORS_WRITE = "vendors.write", _("Реестр поставщиков, изменение")

    CONTRACTS_READ = "contracts.read", _("Контракты, чтение")
    CONTRACTS_WRITE = "contracts.write", _("Контракты, изменение")

    TARIFFS_READ = "tariffs.read", _("Тарифы клиентов, чтение")
    TARIFFS_WRITE = "tariffs.write", _("Тарифы клиентов, изменение")

    BILLING_READ = "billing.read", _("Котировки и счета, чтение")
    BILLING_WRITE = "billing.write", _("Котировки и счета, изменение")

    PAYABLES_READ = "payables.read", _("Заявки на оплату, чтение")
    PAYABLES_WRITE = "payables.write", _("Заявки на оплату, изменение")

    RECONCILIATION_WRITE = "reconciliation.write", _("Сверка счетов")

    REPORTS_DISPATCHER = "reports.dispatcher", _("Дашборд диспетчера")
    REPORTS_FINANCE = "reports.finance", _("Финансовые отчёты")
    REPORTS_VENDOR = "reports.vendor", _("Отчёты по своим услугам")
    REPORTS_ALL = "reports.all", _("Все отчёты")

    AUDIT_READ = "audit.read", _("Журнал действий")

    USERS_MANAGE = "users.manage", _("Управление пользователями")
    SETTINGS_WRITE = "settings.write", _("Настройки системы")
    INTEGRATIONS_MANAGE = "integrations.manage", _("Внешние подключения")


P = Permission

# Матрица из `SPEC.md § 2.2`, строка в строку. Значение «только свои»
# превращается в выданное право плюс фильтрацию выборки, значение
# «заявка» — в отдельное право на подачу заявки.
ROLE_PERMISSIONS: dict[str, frozenset[str]] = {
    Role.ADMIN: frozenset(Permission.values),
    Role.DISPATCHER: frozenset({
        P.FLIGHTS_READ, P.FLIGHTS_WRITE, P.FLIGHTS_STATUS,
        P.FLIGHT_REQUESTS_REVIEW,
        P.ORDERS_CREATE, P.ORDERS_CONFIRM,
        P.CATALOG_READ,
        P.VENDORS_READ,
        P.CONTRACTS_READ,
        P.BILLING_READ,
        P.REPORTS_DISPATCHER,
    }),
    Role.FINANCE: frozenset({
        P.FLIGHTS_READ,
        P.CATALOG_READ, P.CATALOG_WRITE,
        P.VENDORS_READ,
        P.CONTRACTS_READ, P.CONTRACTS_WRITE,
        P.TARIFFS_READ, P.TARIFFS_WRITE,
        P.BILLING_READ, P.BILLING_WRITE,
        P.PAYABLES_READ, P.PAYABLES_WRITE,
        P.RECONCILIATION_WRITE,
        P.REPORTS_FINANCE,
    }),
    Role.MANAGER: frozenset({
        P.FLIGHTS_READ,
        P.CATALOG_READ,
        P.VENDORS_READ,
        P.CONTRACTS_READ,
        P.TARIFFS_READ,
        P.BILLING_READ,
        P.PAYABLES_READ,
        P.REPORTS_ALL, P.REPORTS_DISPATCHER, P.REPORTS_FINANCE,
        P.AUDIT_READ,
    }),
    # ADR-011, требует подтверждения заказчика (G-06). Роль объявлена
    # в карте прав, но пользователи с ней не создаются до ответа.
    Role.SALES: frozenset({
        P.FLIGHTS_READ,
        P.CATALOG_READ,
        P.TARIFFS_READ,
        P.BILLING_READ, P.BILLING_WRITE,
        P.REPORTS_FINANCE,
    }),
    # Порталы. Право выдано, видимость ограничена фильтрацией выборки.
    Role.CLIENT: frozenset({
        P.FLIGHTS_READ,
        P.FLIGHT_REQUESTS_CREATE,
        P.ORDERS_REQUEST,
        P.TARIFFS_READ,
        P.BILLING_READ,
    }),
    Role.VENDOR: frozenset({
        P.ORDERS_CONFIRM,
        P.CATALOG_READ,
        P.VENDORS_READ,
        P.CONTRACTS_READ,
        P.PAYABLES_READ,
        P.REPORTS_VENDOR,
    }),
}

# Роли, ограниченные своими записями (`BACKEND.md § 3.7`). Обращение
# к чужой записи даёт 404, а не 403: существование чужой записи не раскрывается.
TENANT_SCOPED_ROLES = frozenset({Role.CLIENT, Role.VENDOR})

# Перечень ролей, для которых второй фактор обязателен, задан настройкой
# TWO_FACTOR_REQUIRED_ROLES (`BACKEND.md § 6`: «настраивается»).


def permissions_for(role: str) -> frozenset[str]:
    return ROLE_PERMISSIONS.get(role, frozenset())


def permission_map(role: str) -> dict[str, bool]:
    """Карта для `/auth/me`: перечислены все права, а не только выданные.

    Клиент должен уметь отличить «права нет» от «право появилось в новой
    версии сервера и клиент о нём не знает».
    """
    granted = permissions_for(role)
    return {value: value in granted for value in Permission.values}


def has_permission(role: str, permission: Permission | str) -> bool:
    return str(permission) in permissions_for(role)
